src/db/car/plain.rs

```diff
@@ -306,8 +306,8 @@ fn cid_error_to_io_error(cid_error: cid::Error) -> io::Error {
 /// ```
 #[tracing::instrument(level = "trace", skip_all, ret)]
 fn read_header(mut reader: impl Read) -> io::Result<CarHeader> {
-    let header_len =
-        read_varint_body_length_or_eof(&mut reader)?.ok_or(io::Error::from(UnexpectedEof))?;
+    let header_len = read_varint_body_length_or_eof(&mut reader)?
+        .ok_or_else(|| io::Error::from(UnexpectedEof))?;
```

Comment:
This is no allocation but the underlying logic is more than constructing a struct

```rust
#[inline]
    pub(super) fn new_simple(kind: ErrorKind) -> Self {
        let utagged = ((kind as usize) << 32) | TAG_SIMPLE;
        // Safety: `TAG_SIMPLE` is not zero, so the result of the `|` is not 0.
        let res = Self(unsafe { NonNull::new_unchecked(ptr::invalid_mut(utagged)) }, PhantomData);
        // quickly smoke-check we encoded the right thing (This generally will
        // only run in std's tests, unless the user uses -Zbuild-std)
        debug_assert!(
            matches!(res.data(), ErrorData::Simple(k) if k == kind),
            "repr(simple) encoding failed {:?}",
            kind,
        );
        res
    }
```